docs/proposals/ktl.md

```diff
@@ -0,0 +1,243 @@
+   - [Graduation Criteria](#graduation-criteria)
+   - [Alternatives](#alternatives-optional)
+   - [Unresolved Questions](#unresolved-questions-optional)
+
+## Summary
+
+- Stop trying to wholesale move kubectl out of kubernetes/kubernetes
+  - we have been working on this for a year with limited success
+- Instead, create a new cli command called `ktl`.
+  - `ktl` lives in its own repo and vendors commands developed in other repos (kubernetes/kubernetes, kubernetes/kubectl)
+- `ktl` is built and released frequently
+  - vendored sub commands follow their own release cycles and updated by released the vendored code
+
+## Motivation
+
+The core Kubernetes cli is published as a single binary called `kubectl`.
+`kubectl` is a statically linked go binary in the kubernetes/kubernetes
+repo that contains many sub commands to perform operations against a Kubernetes cluster, such as:
+
+- creating & updating objects defined by commandline args
+- managing objects using configuration files
+- debugging objects in a cluster
+
+Benefits of this approach:
+
+- having a single binary facilitates discovery for the suite of Kubernetes cli commands
+- static binary makes it simple to distribute the cli across all OS platforms
+- static binary makes it simple for users to install the cli
+- single binary limits the size of the install (MB) vs distributing ~50 separate go commands (GB)
+- centralizing the development facilitates the construction of shared libraries and infrastructure
+- leveraging the kubernetes/kubernetes repo infrastructure provides a process to build and release
+
+Challenges of this approach:
+
+- the cli cannot be released at a different intervals from the cluster itself
+  - slows down velocity
+  - slows down feedback on alpha and beta features - longer time to GA
+  - each release is heavier weight due to its length
+  - obfuscates version skew as a first class problem
+- individual commands cannot be released at different intervals from one another
+  - some commands should be released every 3 months with the kubernetes cluster
+  - others could be released daily
+- shared cli infrastructure cannot easily be publish to be used by repos / commands outside kubernetes/kubernetes
+  - requires vendoring kubernetes/kubernetes which is painful
+- cli cannnot be owned and maintained independent from unrelated components
+  - submit / merge queue blocks on unrelated tests
+  - GitHub permissions apply to whole repo - cannot add collaborators or maintainers for just the cli code
+  - GitHub notifications for kubernetes/kubernetes are a firehose
+  - hard to manage PRs and issues because they are not scoped to CLI
+
+To address these challenges, sig-cli maintainers have worked toward moving kubectl out of the main repo
+for the past year.  This process has been slow and challenging due to the following reasons:
+
+- many kubectl commands depend on internal kubernetes/kubernetes libraries
```

Comment:
If you vendor current kubectl you'll end up having the same problem, still. It won't magically disappear just by vendoring. In a lot of cases you'll still have to handle that.

Reply:
We should switch to vendoring separate Kubernetes repos, like `kubernetes/client-go`